File: src/shared_memory.cpp
#include "shared_memory.h"

SharedMemory::SharedMemory(const char* name, size_t size) : name(name), size(size) {
    shm_fd = shm_open(name, O_CREAT | O_RDWR, 0666);
    if (shm_fd == -1) {
        std::cerr << "Shared memory oluşturulamadı!\n";
        exit(1);
    }
    ftruncate(shm_fd, size);
    ptr = mmap(0, size, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
}

SharedMemory::~SharedMemory() {
    munmap(ptr, size);
    close(shm_fd);
    shm_unlink(name);
}

void SharedMemory::write(const void* data, size_t size) {
    memcpy(ptr, data, size);
}

void SharedMemory::read(void* buffer, size_t size) {
    memcpy(buffer, ptr, size);
}

File: ai/shared_memory.py
import mmap
import numpy as np
import os

SHM_NAME = "/gameoflife_shm"
WIDTH = 20
HEIGHT = 10
SHM_SIZE = WIDTH * HEIGHT * 4  # int başına 4 byte

class SharedMemory:
    def __init__(self, name, size):
        self.name = name
        self.size = size
        self.fd = os.open(name, os.O_RDWR)
        self.mm = mmap.mmap(self.fd, size, mmap.MAP_SHARED, mmap.PROT_READ | mmap.PROT_WRITE)

    def read(self):
        self.mm.seek(0)
        data = np.frombuffer(self.mm.read(self.size), dtype=np.int32)
        return data.reshape((HEIGHT, WIDTH))

    def write(self, data):
        self.mm.seek(0)
        self.mm.write(data.astype(np.int32).tobytes())

    def close(self):
        self.mm.close()
        os.close(self.fd)

if __name__ == "__main__":
    shm = SharedMemory(SHM_NAME, SHM_SIZE)
    while True:
        grid = shm.read()
        print("\n".join("".join("█" if cell else " " for cell in row) for row in grid))
        print("\n" + "="*40)
